### PR TITLE
fix dataset.reduce issue for optimization directory

### PR DESCRIPTION
**Pull request by hongliny**

### CODE REVIEW

**Comment by hongliny** [comment]
@googlebot I signed it!

**Review by zcharles8 — APPROVED** [review]
This looks great, thanks Honglin! I'll work on getting this pushed internally.

**Comment by zcharles8** [comment]
While this PR looks good to me, we noticed some strange behavior of Keras models with dropout when we changed to `from batch in iter(dataset)`. This is purely a TF issue (involving unexpected interactions between TF dataset ops and Keras layers).

We are currently investigating this, and will push this PR when we have figured out what's going on.

**Comment by hbmcmahan** [comment]
Quick update on this --- we are still following up with the TensorFlow team to see if we can identify the underlying issue with Keras.

**Comment by amitport** [comment]
@hbmcmahan hi, can you expand on the issue? Should we currently use the experimental loop with nightly versions of tf and tff?

**Comment by zcharles8** [comment]
@amitport Just to clarify: You refer to `use_experimental_loop`, which I take it means that you are using `tff.learning`? Just want to understand the scope here, since it doesn't sound like this is part of the PR in question here.

**Comment by amitport** [comment]
I mean both. If I'm using this PR for running with multiple GPUs or using tff.learning `use_experimental_loop` (which seems to be doing the same thing internally), am I'm getting inaccurate results? If tff.learning is OK while this PR isn't, I wonder what is the difference between these implemtations (of the dataset iteration)
Thank you!

**Comment by zcharles8** [comment]
@amitport @hongliny Our current recommendation is to avoid using `iter(dataset)` or `use_experimental_loop` unless you are doing multi-GPU experiments, in which case it is necessary.

In most cases, we have not seen any differences between training behavior in either case. The only cases I have seen are in models that use dropout layers. I don't have any idea what's happening there. It seems like training still occurs using `dataset` or `iter(dataset)`, but the model updates differ slightly. I don't have any more details yet unfortunately, this is a pretty mysterious thing.

**Comment by zcharles8** [comment]
One last clarification: This weird behavior of dropout when using `iter(dataset)` is a pure TF issue, and occurs independently of TFF, so I'm not sure we have much ability to fix it.

**Comment by amitport** [comment]
@zcharles8 OK thanks! To be on the safe side, where reproducibility matters, I wound be using multi-GPU for a while.
